### PR TITLE
Moved sync pod on Node logic to func.

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <madaxa@cn.ibm.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #59194

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/controller/daemon/daemon_controller.go [hunk 1]
```diff
@@ -801,6 +801,68 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
...
+// syncPodOnNode sync the DaemonSet Pods on the node, the return values are:
+//   - nodesNeedingDaemonPods: the pods need to start on the node
+//   - podsToDelete: the Pods need to be deleted on the node
+func (dsc *DaemonSetsController) syncPodOnNode(
+	node *v1.Node,
+	nodeToDaemonPods map[string][]*v1.Pod,
+	ds *extensions.DaemonSet,
+) (nodesNeedingDaemonPods, podsToDelete []string, failedPodsObserved int, err error) {
+
+	wantToRun, shouldSchedule, shouldContinueRunning, err := dsc.nodeShouldRunDaemonPod(node, ds)
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: I assume after all your changes are done, we will only get `wantToRun` and we will eliminate `shouldSchedule`. Basically, DS controller won't check whether node can run the Pod and leaves that decision to the scheduler.

Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: yes. The target is to remove it, I think we will start this feature from alpha, so maybe we will keep it as is for 1.10.

Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller.go]: SGTM

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -801,6 +801,68 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
 	return ds
 }
 
+// syncPodOnNode sync the DaemonSet Pods on the node, the return values are:
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: s/sync/syncs/

The function returns four values. Only two are described in the comment.

**Review by bsalamat — COMMENTED** [review]
LGTM. Just a minor fix for the comment.

File: pkg/controller/daemon/daemon_controller.go (hunk 2 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done

File: pkg/controller/daemon/daemon_controller.go [hunk 3]
```diff
@@ -821,55 +885,20 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet, hash string) e
 	var nodesNeedingDaemonPods, podsToDelete []string
 	var failedPodsObserved int
 	for _, node := range nodeList {
-		wantToRun, shouldSchedule, shouldContinueRunning, err := dsc.nodeShouldRunDaemonPod(node, ds)
+		var nodesNeedingDaemonPodsOnNode, podsToDeleteOnNode []string
+		var failedPodsObservedOnNode int
+		var err error
+
+		nodesNeedingDaemonPodsOnNode, podsToDeleteOnNode, failedPodsObservedOnNode, err = dsc.syncPodOnNode(
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Just use `:=` here

File: pkg/controller/daemon/daemon_controller.go [hunk 4]
```diff
@@ -801,6 +801,70 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
...
 
+// syncPodOnNode syncs the DaemonSet Pods on the node, the return values are:
+//   - nodesNeedingDaemonPods: the pods need to start on the node
+//   - podsToDelete: the Pods need to be deleted on the node
+//   - failedPodsObserved: the number of failed pods on node
+//   - err: unexpected error
+func (dsc *DaemonSetsController) syncPodOnNode(
+	node *v1.Node,
+	nodeToDaemonPods map[string][]*v1.Pod,
+	ds *extensions.DaemonSet,
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: You only need to pass `node.Name`, `nodeToDaemonPods[node.Name]` and `dsKey`

File: pkg/controller/daemon/daemon_controller.go [hunk 5]
```diff
@@ -801,6 +801,70 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
 	return ds
 }
 
+// syncPodOnNode syncs the DaemonSet Pods on the node, the return values are:
+//   - nodesNeedingDaemonPods: the pods need to start on the node
+//   - podsToDelete: the Pods need to be deleted on the node
+//   - failedPodsObserved: the number of failed pods on node
+//   - err: unexpected error
+func (dsc *DaemonSetsController) syncPodOnNode(
```
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: nit: Need to rename this function. "Sync" means reconciles the current state to make it match the given desired state (e.g. `syncNodes` in this file syncs by creating/deleting pods on nodes). 

What this function does is calculating the diff. Suggest naming it something like "podsShouldOnNode". The comment can be something like "podsShouldOnNode figures out the DaemonSet pods to be created and deleted on the given node".

File: pkg/controller/daemon/daemon_controller.go (hunk 3 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done

File: pkg/controller/daemon/daemon_controller.go (hunk 4 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: seems not change on # of arg, so I keep it :)

File: pkg/controller/daemon/daemon_controller.go (hunk 5 above)
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done :)

**Review by janetkuo — APPROVED** [review]

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/59636)

File: pkg/controller/daemon/daemon_controller.go [hunk 6]
```diff
@@ -802,6 +802,70 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
 	return ds
 }
 
+// podsShouldOnNode figures out the DaemonSet pods to be created and deleted on the given node:
+//   - nodesNeedingDaemonPods: the pods need to start on the node
+//   - podsToDelete: the Pods need to be deleted on the node
+//   - failedPodsObserved: the number of failed pods on node
+//   - err: unexpected error
+func (dsc *DaemonSetsController) podsShouldOnNode(
```
Comment (mikedanese) [inline pkg/controller/daemon/daemon_controller.go]: grammar nit: podsShouldOnNode needs a verb.

**Comment by mikedanese** [comment]
I'll let @bsalamat readd the lgtm.

**Pull request merged**